test/serviceprofiles/serviceprofiles_test.go

```diff
@@ -0,0 +1,240 @@
+package serviceprofiles
+
+import (
+	"fmt"
+	"os"
+	"strings"
+	"testing"
+
+	"github.com/cloudflare/cfssl/log"
+	"github.com/linkerd/linkerd2/testutil"
+)
+
+var TestHelper *testutil.TestHelper
+
+type testCase struct {
+	namespace   string
+	injectYAML  string
+	deployments []string
+	deployName  string
+	spName      string
+}
+
+func TestMain(m *testing.M) {
+	TestHelper = testutil.NewTestHelper()
+	code := m.Run()
+	tearDown()
+	os.Exit(code)
+}
+
+func TestServiceProfilesFromTap(t *testing.T) {
+	testCases := []testCase{
+		{
+			namespace:   "emojivoto",
+			injectYAML:  "emojivoto.yml",
+			deployments: []string{"emoji", "vote-bot", "voting", "web"},
+			deployName:  "deploy/voting",
+			spName:      "voting-svc",
+		},
+		{
+			namespace:   "booksapp",
+			injectYAML:  "booksapp.yml",
+			deployments: []string{"webapp", "authors", "books", "traffic"},
+			deployName:  "deploy/books",
+			spName:      "books-svc",
+		},
+	}
+
+	for _, tc := range testCases {
+		t.Run(fmt.Sprintf("service profiles from tap:%s", tc.namespace), func(t *testing.T) {
+			cmd := []string{"inject", fmt.Sprintf("testdata/%s", tc.injectYAML)}
+			out, _, err := TestHelper.LinkerdRun(cmd...)
+			if err != nil {
+				t.Fatalf("linkerd inject command failed: %s\n", err)
+			}
+
+			out, err = TestHelper.KubectlApply(out, tc.namespace)
+			if err != nil {
+				t.Fatalf("kubectl apply command failed:\n%s", out)
+			}
+
+			for _, deploy := range tc.deployments {
+				err = TestHelper.CheckPods(tc.namespace, deploy, 1)
+				if err != nil {
+					t.Fatalf("Unexpected error: %s\n", err.Error())
+				}
+			}
+
+			// run routes: Expect default route only
+			routes, err := getRoutes(tc.deployName, tc.namespace, TestHelper)
+			if err != nil {
+				t.Fatalf("routes command failed: %s\n", err)
+			}
+
+			if len(routes) > 1 {
+				t.Fatalf("Expected route details for service to be at most 1 but got %d\n", len(routes))
+			}
+
+			// run service profile from tap command
+			cmd = []string{
+				"profile",
+				"--namespace",
+				tc.namespace,
+				tc.spName,
+				"--tap",
+				tc.deployName,
+				"--tap-route-limit",
+				"5",
+				"--tap-duration",
+				"10s"}
+			out, _, err = TestHelper.LinkerdRun(cmd...)
+			if err != nil {
+				t.Fatalf("profile command failed: %s\n", err.Error())
+			}
+
+			out, err = TestHelper.KubectlApply(out, tc.namespace)
+			if err != nil {
+				t.Fatalf("kubectl apply command failed:\n%s", out)
+			}
+
+			// run routes: Expected more than default route
+			routes, err = getRoutes(tc.deployName, tc.namespace, TestHelper)
+			if err != nil {
+				t.Fatalf("routes command failed: %s\n", err.Error())
+			}
+			for _, route := range routes {
+				if len(route) <= 1 {
```

Comment:
i think `len(route)` is testing the length of each route name? should this instead be...
```go
if len(routes) <= 1 {
	t.Fatalf("Expected routes for service to be greater than or equal to 1 but got %d\n", len(routes))
}
```